### PR TITLE
IGNITE-15188 JDBC driver for 3.0: Database metadata

### PR DESCRIPTION
**Pull request by vladErmakov07**
https://issues.apache.org/jira/browse/IGNITE-15188

### CODE REVIEW

File: modules/client-common/src/main/java/org/apache/ignite/client/proto/query/event/JdbcMetaTablesResult.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+            meta = new ArrayList<>(size);
+
+            for (int i = 0; i < size; ++i) {
+                JdbcTableMeta m = new JdbcTableMeta();
+
+                m.readBinary(unpacker);
+
+                meta.add(m);
+            }
+        }
```
Comment (AMashenkov) [inline modules/client-common/src/main/java/org/apache/ignite/client/proto/query/event/JdbcMetaTablesResult.java:88]:
```suggestion
        meta = new ArrayList<>(size);

        for (int i = 0; i < size; ++i) {
            JdbcTableMeta m = new JdbcTableMeta();

            m.readBinary(unpacker);

            meta.add(m);
        }
```

File: modules/client-common/src/main/java/org/apache/ignite/client/proto/query/event/JdbcTableMeta.java [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        JdbcTableMeta meta = (JdbcTableMeta)o;
+
+        return Objects.equals(schemaName, meta.schemaName)
+            && Objects.equals(tblName, meta.tblName)
+            && Objects.equals(tblType, meta.tblType);
+    }
+
+    /** {@inheritDoc} */
+    @Override public int hashCode() {
+        return Objects.hash(schemaName, tblName, tblType);
```
Comment (AMashenkov) [inline modules/client-common/src/main/java/org/apache/ignite/client/proto/query/event/JdbcTableMeta.java:128]: 1. Implicit array instance.
2. inconsistent with JdbcPrimaryKeyMeta

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/sql/JdbcMetadataInfo.java [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+import org.apache.ignite.internal.schema.NativeTypeSpec;
+import org.apache.ignite.internal.schema.NumberNativeType;
+import org.apache.ignite.internal.schema.SchemaDescriptor;
+import org.apache.ignite.internal.schema.SchemaRegistry;
+import org.apache.ignite.internal.table.TableImpl;
+import org.apache.ignite.internal.util.Pair;
+import org.apache.ignite.table.Table;
+import org.apache.ignite.table.manager.IgniteTables;
+
+//TODO Filter by table type must be added after 'view' type will appear.
```
Comment (AMashenkov) [inline modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/sql/JdbcMetadataInfo.java:48]: 'TODO' without ticket number.

File: modules/client/src/main/java/org/apache/ignite/internal/jdbc/JdbcDatabaseMetadata.java [hunk 4]
```diff
@@ -0,0 +1,1529 @@
...
+    @Override public ResultSet getProcedures(String catalog, String schemaPtrn,
+        String procedureNamePtrn) {
+        return new JdbcResultSet(Collections.emptyList(), asList(
+            new JdbcColumnMeta(null, null, "PROCEDURE_CAT", String.class),
+            new JdbcColumnMeta(null, null, "PROCEDURE_SCHEM", String.class),
+            new JdbcColumnMeta(null, null, "PROCEDURE_NAME", String.class),
+            new JdbcColumnMeta(null, null, "REMARKS", String.class),
+            new JdbcColumnMeta(null, null, "PROCEDURE_TYPE", String.class),
+            new JdbcColumnMeta(null, null, "SPECIFIC_NAME", String.class)
+        ));
```
Comment (AMashenkov) [inline modules/client/src/main/java/org/apache/ignite/internal/jdbc/JdbcDatabaseMetadata.java:678]: Should we throw SqlException(UnsupportedOperationException) here as log as supportsStoredProcedures() returns false?

File: modules/client/src/main/java/org/apache/ignite/internal/jdbc/JdbcDatabaseMetadata.java [hunk 5]
```diff
@@ -0,0 +1,1529 @@
...
+            new JdbcColumnMeta(null, null, "NULLABLE", Short.class),
+            new JdbcColumnMeta(null, null, "REMARKS", String.class),
+            new JdbcColumnMeta(null, null, "COLUMN_DEF", String.class),
+            new JdbcColumnMeta(null, null, "SQL_DATA_TYPE", Integer.class),
+            new JdbcColumnMeta(null, null, "SQL_DATETIME_SUB", Integer.class),
+            new JdbcColumnMeta(null, null, "CHAR_OCTET_LENGTH", Integer.class),
+            new JdbcColumnMeta(null, null, "ORDINAL_POSITION", Integer.class),
+            new JdbcColumnMeta(null, null, "IS_NULLABLE", String.class),
+            new JdbcColumnMeta(null, null, "SPECIFIC_NAME", String.class)
+            ));
```
Comment (AMashenkov) [inline modules/client/src/main/java/org/apache/ignite/internal/jdbc/JdbcDatabaseMetadata.java:705]: Stored procedures are not supported

File: modules/client/src/main/java/org/apache/ignite/internal/jdbc/JdbcDatabaseMetadata.java [hunk 6]
```diff
@@ -0,0 +1,1529 @@
...
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean supportsExtendedSQLGrammar() {
+        return false;
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean supportsANSI92EntryLevelSQL() {
+        return true;
```
Comment (AMashenkov) [inline modules/client/src/main/java/org/apache/ignite/internal/jdbc/JdbcDatabaseMetadata.java:345]: I'm not sure =)

File: modules/client/src/main/java/org/apache/ignite/internal/jdbc/JdbcDatabaseMetadata.java [hunk 7]
```diff
@@ -0,0 +1,1529 @@
...
+        return new JdbcResultSet(Collections.emptyList(), asList(
+            new JdbcColumnMeta(null, null, "TABLE_CAT", String.class),
+            new JdbcColumnMeta(null, null, "TABLE_SCHEM", String.class),
+            new JdbcColumnMeta(null, null, "TABLE_NAME", String.class),
+            new JdbcColumnMeta(null, null, "GRANTOR", String.class),
+            new JdbcColumnMeta(null, null, "GRANTEE", String.class),
+            new JdbcColumnMeta(null, null, "PRIVILEGE", String.class),
+            new JdbcColumnMeta(null, null, "IS_GRANTABLE", String.class)
+        ));
+    }
```
Comment (AMashenkov) [inline modules/client/src/main/java/org/apache/ignite/internal/jdbc/JdbcDatabaseMetadata.java:850]: Privileges are not supported.

File: modules/client/src/main/java/org/apache/ignite/internal/jdbc/JdbcDatabaseMetadata.java [hunk 8]
```diff
@@ -0,0 +1,1529 @@
...
+            new JdbcColumnMeta(null, null, "FK_NAME", String.class),
+            new JdbcColumnMeta(null, null, "PK_NAME", String.class),
+            new JdbcColumnMeta(null, null, "DEFERRABILITY", Short.class)
+        ));
+    }
+
+    /** {@inheritDoc} */
+    @Override public ResultSet getCrossReference(String parentCatalog, String parentSchema, String parentTbl,
+        String foreignCatalog, String foreignSchema, String foreignTbl) {
+        return new JdbcResultSet(Collections.emptyList(), asList(
```
Comment (AMashenkov) [inline modules/client/src/main/java/org/apache/ignite/internal/jdbc/JdbcDatabaseMetadata.java:952]: Foreign constraints are not supported.

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/sql/JdbcMetadataInfo.java [hunk 9]
```diff
@@ -0,0 +1,329 @@
...
+/**
+ * Facade over {@link IgniteTables} to get information about database entities in terms of JDBC.
+ */
+public class JdbcMetadataInfo {
```
[the remark this answers is not in the document]
Comment (AMashenkov) [reply modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/sql/JdbcMetadataInfo.java:52]: or JdbcMetadataService

File: modules/client-common/src/main/java/org/apache/ignite/client/proto/query/event/JdbcTableMeta.java (hunk 2 above)
Comment (vladErmakov07) [reply modules/client-common/src/main/java/org/apache/ignite/client/proto/query/event/JdbcTableMeta.java]: Rewrited hashCode, thx.

File: modules/client/src/main/java/org/apache/ignite/internal/jdbc/JdbcDatabaseMetadata.java (hunk 6 above)
Comment (vladErmakov07) [reply modules/client/src/main/java/org/apache/ignite/internal/jdbc/JdbcDatabaseMetadata.java:346]: I added separate task to check it out.

File: modules/client/src/main/java/org/apache/ignite/internal/jdbc/JdbcDatabaseMetadata.java [hunk 10]
```diff
@@ -0,0 +1,1529 @@
...
+        if (!res.hasResults())
+            throw IgniteQueryErrorCode.createJdbcSqlException(res.err(), res.status());
+
+        List<List<Object>> rows = new LinkedList<>();
+
+        for (String schema : res.schemas()) {
+            List<Object> row = new ArrayList<>(2);
+
+            row.add(schema);
+            row.add(CATALOG_NAME);
```
[the remark this answers is not in the document]
Comment (vladErmakov07) [reply modules/client/src/main/java/org/apache/ignite/internal/jdbc/JdbcDatabaseMetadata.java:1331]: I added PUBLIC as default schema in JdbcMetadataCatalog#getSchemasMeta

**Review by AMashenkov — APPROVED** [review]

**Pull request merged**